### PR TITLE
Add dicom visual navigation

### PR DESCRIPTION
**Pull request by jcfr**

### CODE REVIEW

**Comment by jcfr** [comment]
@Punzo While reviewing the changes I ended up creating a pull request and marked it as draft.

**Comment by Punzo** [comment]
> @Punzo While reviewing the changes I ended up creating a pull request and marked it as draft.

no problem!

btw: I am still actively working on it.

Here it is my current To Do list:

- add a storage listener class in CTK (ctkDICOMStorageListener). At the moment the visual browser works only with C-GET.
- implement send to server in right click context menu (at the moment it just give a signal with the files list). It would be nice to have this, to use the taskPool manager for sending to server (UI responsive, etc....).
- integrate the widget in Slicer. For the moment the widget should be an expertimental option. For example, we could add a query and retrieve experimental button in the Slicer UI which would open the visual dicom browser.
- Add dicomweb query and retrieve (We would need to implement a new C++ library).
- Implement streaming to volume nodes. i.e. no need to wait the series to be fully donwloaded to import them.
- drag and drop of a DICOM series widget thumbnail on a Slicer view to import it automatically into a volume node and activate the volumeNode in the view.

**Comment by jcfr** [comment]
> Add dicomweb query and retrieve (We would need to implement a new C++ library).

cc: @tbirdso

**Comment by lassoan** [comment]
Ideally, DICOMweb query/retrieve would use a DICOM toolkit, such as DCMTK, but unfortunately it does not seem like DCMTK will ever have DICOMweb networking. We could look for other C++ implementations, or implement using CTK's qRestAPI, or worst case call a Python implementation (dicomweb-client) from C++.

**Comment by Punzo** [comment]
> Ideally, DICOMweb query/retrieve would use a DICOM toolkit, such as DCMTK, but unfortunately it does not seem like DCMTK will ever have DICOMweb networking. We could look for other C++ implementations, or implement using CTK's qRestAPI, or worst case call a Python implementation (dicomweb-client) from C++.

for future reference:
https://github.com/MITK/MITK/tree/master/Modules/DICOMweb

**Comment by Punzo** [comment]
> FYI @Punzo I'm investigating DICOMweb loading too, so we can touch base to compare notes at some point. Still working on some experimental code.

Ciao Steve, ok that would be great! 

I did not have the time to investigate DICOMweb in C++ yet and I will focus next days on the first 4 points in my To Do [list](https://github.com/commontk/CTK/pull/1142#issuecomment-1745083628), but I would be glad to connect soon. Let's update soon.

**Comment by Punzo** [comment]
> > @Punzo While reviewing the changes I ended up creating a pull request and marked it as draft.
> 
> no problem!
> 
> btw: I am still actively working on it.
> 
> Here it is my current To Do list:
> 
> * [x]  add a storage listener class in CTK (ctkDICOMStorageListener).
> * [x]  binaries in CTK with an example of the visual dicom browser running without Slicer (Folder selector for the database and ctkDICOMVisualBrowserWidget).
> * [x]  PR ready for review

@jcfr @pieper @lassoan this PR is ready for a first review 
pinging also @nolden.

The operations are executed by a task pool manager in separate threads with a priority queue.
The current supported operations are:

- Filtering and navigation with thumbnails of local database and servers results
- Import from file system to local database
- Query/Retrieve from servers (DIMSE C-GET/C-MOVE )
- Storage listener
- Send (emits only a signal for the moment, requires external implementation)
- Remove (only from local database, not from server)
- Metadata exploration

UML diagram:
image

video:

https://github.com/commontk/CTK/assets/7985338/f46c1c90-6b9e-493a-b3e5-63634062911b

I have added a ctkApplication called ctkDICOMVisualBrowser for testing (need to set cmake options `examples` and `DICOM` on).

Future enh can be:
- Integrate in Slicer (probably as an additional experimental option in the DICOM module)
- Clean the UML design. At the moment the classes are too dependent on each other, since the updates are done by signal with raw pointers (`ctkDICOMTaskResults`). Moreover in this way the role of deleting of the tasks is on the UI and we need to wait the UI updates. Probably it would be better to close the task as soon as possible (e.g. for retrieve, add the inserter operation directly in the task and delete the task as soon as finish). Then UI updates should be done sending simple strings (PatientID, studyInstanceUID, etc...) and the UI takes the data for updating directly from the local dicom database. This would optimize and make the the memory managment clean and simple.
- Implementing send (i.e. adding `ctkDICOMSendTask` and `ctkDICOMSend` with underlining DIMSE `DcmStorageSCU`)
- add DICOMweb

**Comment by Punzo** [comment]
> ## To Do: cleaning design
> At the moment the classes are too dependent on each other:
> 
> * the updates are done by sending signals with raw pointers `ctkDICOMTaskResults` (queued connections for slots). These raw pointers are created and owned by the Tasks and the data are needed both from the indexer and the UI.
> * the role of deleting the tasks is on the UI.
> * we need to wait the UI updates before deleting the tasks.
> 
> An improved design with a clean and simple memory managment should:
> 
> * destroying the task as soon as possible -> e.g. for retrieve, add the inserter operation directly in the task and delete the task as soon as the indexing is finished.
> * UI updates should be done by sending simple strings (PatientID, studyInstanceUID, etc...) from the tasks to the TaskPool and from the TaskPool to the UI (direct connections for slots).
> * the UI access the data for the update directly from the local dicom database using the information strings (PatientID, studyInstanceUID, etc...).
> * task dependencies (e.g., retrieve with C-MOVE / C-GET with a proxy server) logic should not be hardcoded in the TaskPool. We need a class that collects the tasks and put them in a queue and prints to a file (JSON?). Then the TaskPool simply run the QRunnables. This would allow also resuming tasks if the UI crash or gets restarted etc....

Done in https://github.com/commontk/CTK/pull/1142/commits/f9119726c3e52628820ffc16e97c71c64252c5b3 
1) Logic and UI are now independent
2) each worker is now independent and it can start new jobs independently, e.g.: retry or proxy server.  `Workers` are the executor, `jobs` are metadata objects to handle the queue.

this is the current UML diagram:
ctkVisualDICOMBrowser

## Current PR content description
The operations are executed by a scheduler in separate threads with a priority queue.
The current supported operations are:

- Filtering and navigation with thumbnails of local database and servers results
- Import from file system to local database
- Query/Retrieve from servers (DIMSE C-GET/C-MOVE ). All the operations are done in background and in parallel
- Storage listener
- Send (emits only a signal for the moment, requires external implementation)
- Remove (only from local database, not from server)
- Metadata exploration
 
## Future ENH:
- Integrate in Slicer (probably as an additional experimental option in the DICOM module)
- Implementing send (i.e. adding `ctkDICOMSendTask` and `ctkDICOMSend` with underlining DIMSE `DcmStorageSCU`)
- add DICOMweb
- handle jobs queue in the scheduler by file (so we can restart the jobs/workers at application restart)

**Comment by Punzo** [comment]
Here's a revised version of your text:

---

@lassoan @pieper @jcfr, this is ready for review. Any feedback on the design of the infrastructure classes, widget, and logic implementation would be greatly appreciated. For a description of the current status, please refer to my previous [comment](https://github.com/commontk/CTK/pull/1142#issuecomment-1845680504).

Please note that I still have a couple of small tasks pending (see below and the "To Do" comments in the code). I plan to address them in January upon my return from vacation, but they are minor enhancements to the operations logic (deep within the code) and should not hinder the review process:

A) Presently, the results from a retrieve operation are inserted at the end of the fetch. This leads to:
   i) Too many inserts for the `RetrieveSOPInstance` operation, resulting in slower processing.
   ii) Temporary saving of too many frames in memory for `RetrieveSeries` and `RetrieveStudy` operations, leading to higher memory consumption, especially for large series.
   iii) For the StorageListener, inserts are done every 1 second for all the fetched frames within that time, causing similar issues as (i) and (ii).

B) Concurrently inserting into the ctkDICOMDatabase from different threads poses an issue where the native lock (transaction + commit) mechanism does not function as expected. I tested this with other databases and encountered this problem only with the scheme in ctkDICOMDatabase. This could potentially lead to crashes (even when opening a database connection for each QThread).

Proposed solutions (requiring only a couple of development days, but I'll address them in January):

(A) For all operations, implement batch insertions whenever the number of fetched frames exceeds a configurable `frameBatchLimit` variable.

(B) Currently, ctkDICOMScheduler ensures that only one ctkDICOMInsertJob is executed at a time by the workers. We should also add a static `Lock` variable in the ctkDICOMDatabase to block any other CTK component or application from writing to the database if another writing process is ongoing.

---------------------------------------------------------------------------------------------------------------------------------------------

Note/question for @jcfr: To run the test locally, I had to follow these steps:
1) I configured CTK in a CTK-Build folder as follows:
```
CTK_BUILD_ALL                    OFF
CTK_BUILD_ALL_APPS               OFF
CTK_BUILD_ALL_LIBRARIES          OFF
CTK_BUILD_ALL_PLUGINS            OFF
CTK_BUILD_EXAMPLES               ON
CTK_ENABLE_DICOM                 ON
CTK_ENABLE_DICOMApplicationHos   OFF
CTK_ENABLE_PluginFramework       OFF
CTK_ENABLE_Python_Wrapping       OFF
CTK_ENABLE_Widgets               ON
```
2) Ran `make -j 20` in CTK-Build
3) Manually ran `make install` in CTK-Build to generate DCMTK binaries in `CTK-Build/CMakeExternals/Install/bin`
4) Added `CTK-Build/CMakeExternals/Install/lib` to my `LD_LIBRARY_PATH` environment variable
5) Then I could run the test in `CTK-Build/CTK-build` folder (`ctest -j 20`)

I might be missing something, but it would be helpful if the superbuild could automatically install DCMTK (eliminating the need for steps 3 and 4). Please let me know if this is a bug or if I am overlooking something.

**Comment by pieper** [comment]
This is quite a lot of code, but a quick readthrough didn't raise any serious concerns for me, although we really should do some cross platform testing with different network and server scenarios to see how all the concurrency holds up.

I think it would help me to have a guided tour of the class structure.  We can do this in January, either on zoom or in person at project week.  Do you have a preferred timeline for merging?

**Comment by lassoan** [comment]
@davide has been doing all development and testing on linux. I'll test on Windows. It would be great if @pieper you could test on macOS.

**Comment by Punzo** [comment]
> This is quite a lot of code, but a quick readthrough didn't raise any serious concerns for me, although we really should do some cross platform testing with different network and server scenarios to see how all the concurrency holds up.
> 

Ok thanks, for having already had a quick look.
I agree, as Andras said I tested on Linux and mainly with Orthanc servers. It would be probably good to test also in environments with low bandwidth and high packet loss.

> I think it would help me to have a guided tour of the class structure.  We can do this in January, either on zoom or in person at project week.  Do you have a preferred timeline for merging?

Sure, we can do a video call asap when we will be back from vacation in January.  Ideally I would like to merge and integrate in Slicer (as an additional experimental feature in the DICOM module) before the project week. In this way we can use the Slicer preview during the slicer project week for testing, collecting feedback, etc.

**Comment by pieper** [comment]
Sounds great.  I'll be on break until Jan 3, but happy to meet after that.  Enjoy your time off : )

**Comment by Punzo** [comment]
> Sounds great.  I'll be on break until Jan 3, but happy to meet after that.  Enjoy your time off : )
> 

Have a nice Christmas vacation too (: !!!!

**Comment by Punzo** [comment]
Note from meeting 5th:
1) Verify host: change response string. C-ECHO connection verified/failed (if failed error)
2) server settings should be more human readeable in the q settings file.
3) Change in the UI the column names Proxy and Protocol with additional Retrieve (Retrieve Proxy etc....)
4) improve query errors, suggest to check the connection settings. Demphasize the request ID, maybe add a detail section where to bump all the info (for example, server name, etc.....)
5) Add debug option to activate the output from DCMTK. Also we should have a look as to stream the DCMTK output. It would be nice also to have the output separated for each request.
6) right click delete name in the UI should be: delete from local database
7) Prefetch the full series (now the current behaviour) could be disabled and the retrieve could it be done only when the selected series are manually loaded. 
8) Thumbnails: add transparent and four directions in the draw text for the thumbnails. Remove N.frames and put the third dimesions on the rowXcolumns (e.g. 100x100x5). Elide the series description and put the full text in tooltip.
9) add number of conccurent workers set per server.
10) retry time delay should be higher and exponential with a factor 2/3 (with some randomness). Current parameter is 100ms (not enough for a server to recover). Maximum number retry should be set to a a maximum waiting time  (defined in the GUI in the server settings) -> timout warning in the UI (check the string if it is clear).

**Comment by lassoan** [comment]
@Punzo Please rebase on latest master. There are some conflicting changes in ctkDICOMDatabase.cxx.

**Comment by Punzo** [comment]
> @Punzo Please rebase on latest master. There are some conflicting changes in ctkDICOMDatabase.cxx.

Done!

File: Libs/DICOM/Core/ctkDICOMWorker.h [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+=========================================================================*/
+
+#ifndef __ctkDICOMWorker_h
+#define __ctkDICOMWorker_h
+
+// Qt includes
+#include <QSharedPointer>
+
+// CTK includes
```
Comment (lassoan) [inline Libs/DICOM/Core/ctkDICOMWorker.h:30]:
```suggestion
// ctkDICOMCore includes
```

File: Libs/DICOM/Core/ctkDICOMWorker.h [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+#include "ctkCoreExport.h"
```
Comment (lassoan) [inline Libs/DICOM/Core/ctkDICOMWorker.h:31]:
```suggestion
#include "ctkDICOMCoreExport.h"
```

File: Libs/DICOM/Core/ctkDICOMWorker.h [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+#include "ctkAbstractWorker.h"
+
+class ctkDICOMJob;
+class ctkDICOMScheduler;
+
+//------------------------------------------------------------------------------
+/// \ingroup Core
+class CTK_CORE_EXPORT ctkDICOMWorker : public ctkAbstractWorker
```
Comment (lassoan) [inline Libs/DICOM/Core/ctkDICOMWorker.h:39]: This fixes a build error on Windows.
 
```suggestion
class CTK_DICOM_CORE_EXPORT ctkDICOMWorker : public ctkAbstractWorker
```

**Comment by lassoan** [comment]
Query/retrieve from dicomserver.co.uk (see server information [here](https://dicomserver.co.uk/)) works well using Slicer's old DICOM networking feature. However, it does not work with the visual DICOM browser.

I've tried query/retrieve from dicomserver.co.uk  and no progress information was shown, no error message was displayed, , nothing showed up in the browser. We really need some feedback about what's going on with the query. https://dicomserver.co.uk/logs/ showed that connection was established and some information was exchanged.

Verification failed, too. A popup was displayed, but would be nicer to have a "Verification" column in the "Servers" table (unknown, in progress, success, failure).

**Comment by lassoan** [comment]
After about half hour of investigation, I've realized that although I've fixed the port number of dicomserver.co.uk, I did not click `Apply changes`. To avoid this, the following could help: if when there are pending changes, then the modified field and the `Apply changes` button could be VERY visibly highlighted.

After Apply button, some patients are retrieved!! Good! However, studies only appeared for a few patients.

**Comment by Punzo** [comment]
> After about half hour of investigation, I've realized that although I've fixed the port number of dicomserver.co.uk, I did not click `Apply changes`. To avoid this, the following could help: if when there are pending changes, then the modified field and the `Apply changes` button could be VERY visibly highlighted.


Ok, I will make this a priority on Monday.

Do you think that setting the text to bold and the cell selected (i.e. different background) would be enough? 

> 
> After Apply button, some patients are retrieved!! Good! However, studies only appeared for a few patients.

Uhm, ok I will have a look on Monday. Can you tell me what is the right port for dicomserver.co.uk? So I will also correct the default.

**Comment by lassoan** [comment]
dicomserver.co.uk - port 104 worked for me:

image

More information on the server and its capabilities:

https://dicomserver.co.uk/

**Comment by lassoan** [comment]
> Do you think that setting the text to bold and the cell selected (i.e. different background) would be enough?

Using the same yellow background as in the filter/query parameter highlighting could work. I don't think bold is necessary.

Few other things:
- Style of "Apply changes" and "Discard changes" should be the same as Add/verify/remove host.
- Error reporting is inconsistent. Most often I don't see any errors (just nothing shows up), but once an error message came up as a button. But then I could not remove that message. A nice solution would be to have a "Job list" button, which could show an error notification (red or yellow circle) on it when one of the jobs failed. Clicking on it the button would show the job list (it can be a popup or a collapsing section in the GUI) that would show all the jobs (in-progress and completed) along with its status (it can be just simple list, details shown as tooltip or popup; with an option to show all jobs or only in-progress jobs, a button to clear the completed jobs, a button to cancel the selected jobs).

image

**Comment by Punzo** [comment]
RoadMap (after item 4, the items are not listed in priority):

- [ ] 1) Apply windows build fix
- [ ] 2) Fix missing studies when querying https://dicomserver.co.uk/ (port 104)
- [ ] 3) merge PR
- [ ] 4) Integrate Visual DICOM Browser in Slicer
- [ ] 5) add a writing lock static variable in the ctkDICOMDatabase class
- [ ] 6) smart managing of the inserts and memory when retrieving a series. i.e. we could have a customizable `framesBatchLimit` variable.
- [ ] 7) Verify host: change response string. C-ECHO connection verified/failed (if failed error)
- [ ] 8) server settings should be more human readeable in the qt settings file.
- [ ] 9) Change in the UI the column names Proxy and Protocol with additional Retrieve (Retrieve Proxy etc....)
- [ ] 10) improve query errors, suggest to check the connection settings. Demphasize the request ID, maybe add a detail section where to bump all the info (for example, server name, etc.....)
- [ ] 11) Add debug option to activate the output from DCMTK. Also we should have a look as to stream the DCMTK output. It would be nice also to have the output separated for each request.
- [ ] 12) right click delete name in the UI should be: delete from local database
- [ ] 13) Prefetch the full series (now the current behaviour) could be disabled and the retrieve could it be done only when the selected series are manually loaded. 
- [ ] 14) Thumbnails: add transparent and four directions in the draw text in the screenshots. Remove N.frames and put the third dimension on the rowXcolumns (e.g. 100x100x5). Elide the series description and put the full text in tooltip.
- [ ] 15) add infrastructure to set the maximum number of concurrent workers per server (now it is just per type).
- [ ] 16) the retry time delay value should be higher and exponential with a factor 2/3 (with some randomness). Current default parameter is a fixed 100ms (not enough for a server to recover). Maximum number of retries should be set to a a maximum waiting time (the one defined in the GUI in the server settings) -> timout warning in the UI (check the string if it is clear).
- [ ] 17) Style of "Apply changes" and "Discard changes" should be the same as Add/verify/remove host. Maybe it would be more clear if Apply/Discard changes buttons would be in each section (to avoid cluttering the GUI, we could make the buttons show up only when there are pending changes)
- [ ] 18) Error reporting is inconsistent. Most often I don't see any errors (just nothing shows up), but once an error message came up as a button. But then I could not remove that message. A nice solution would be to have a "Job list" button, which could show an error notification (red or yellow circle) on it when one of the jobs failed. Clicking on it the button would show the job list (it can be a popup or a collapsing section in the GUI) that would show all the jobs (in-progress and completed) along with its status (it can be just simple list, details shown as tooltip or popup; with an option to show all jobs or only in-progress jobs, a button to clear the completed jobs, a button to cancel the selected jobs, and a button to retry the selected jobs).
- [ ] 19) In the ctkDICOMVisualBrowser application, the folder selector does not reflect the actual DICOM database location (it seems that the button is not updated from application settings at startup)
- [ ] 20) When hitting the search button, the patient list always jumps to the first patient, which is quite annoying, as I keep clicking it to see if new studies have become available for the current patient on the server. Current patient selection should not be lost if Search button is pressed (if the current patient is still selected by the search criteria).
- [ ] 21) By default show all local content.
- [ ] 22) Error message like "bc2343432-234241frfd-2342341s job failed to fetch data" is quite annoying. It should tell something like "Fetching data for patient John Doe [patientid] from server MedicalConnections failed. Response not received for 30 seconds. Click here to see more details." (clicking could open the job list with that task highlighted)
- [ ] 23) In settings, servers should be in a "Servers" collapsible section, the same way as "Storage" (instead of the "Servers" label)
- [ ] 24) Display patient name in "Patient information" section (while there are many patient names listed above, it is not that clear which one is actually selected; it is also not possible to select and copy-paste that name). Left-align field names and values (for example, "ID:" label is on the left - which is good; but then the actual ID value is floating somewhere in the middle of the screen).

**Comment by Punzo** [comment]
@lassoan (1) and (2) done. Please have a look if this can be merged now. I will start (4) meanwhile

**Comment by lassoan** [comment]
Based on review with @pieper and @lassoan and testing on Linux and Windows on multiple servers, this is now ready to be merged.

**Comment by lassoan** [comment]
This is now merged. @Punzo thank you very much for your awesome work!

Please add the list of issues/enhancements to a new issue in this repository.

**Comment by Punzo** [comment]
> * implementing send (i.e. adding `ctkDICOMSendJob`, `ctkDICOMSendWorker` and `ctkDICOMSend` with underlining DIMSE `DcmStorageSCU`)
> * add DICOMweb
> * handle jobs queue in the scheduler by file (so we can restart the jobs/workers at application restart). We should add UI to visualize the jobs queue and status.
> * add data streaming from visual brower series widgets to Slicer volume nodes.

Thanks for reviewing/testing and merging!

**Comment by Punzo** [comment]
> Please add the list of issues/enhancements to a new issue in this repository.

Done, see https://github.com/commontk/CTK/issues/1162

**Comment by jcfr** [comment]
I suggest we revert `CTK` master to the hash before the integration and use `squash & merge` along with providing a descriptive commit message outlining the feature integrated, currently there is only the title `ENH: Add visual DICOM browser`

**Comment by lassoan** [comment]
I was thinking about this today, too. There is a follow-up PR that I'll test and if that works well then we should squash that with all these commits.

**Pull request merged**